libraries/GCS_MAVLink/GCS.h

```diff
@@ -470,11 +465,80 @@ class GCS_MAVLINK
     char _perf_packet_name[16];
     char _perf_update_name[16];
 
-    // deferred message handling.  We size the deferred_message
-    // ringbuffer so we can defer every message type
-    enum ap_message deferred_messages[MSG_LAST];
-    uint8_t next_deferred_message;
-    uint8_t num_deferred_messages;
+    // outbound ("deferred message") queue.
+
+    // "special" messages such as heartbeat, next_param etc are
+    // stored separately to stream-rated messages like AHRS2 etc:
+    struct deferred_message_t {
+        const ap_message id;
+        uint16_t interval_ms;
+        uint16_t last_sent_ms; // from AP_HAL::millis16()
+    } deferred_message[2] = {
+        { MSG_HEARTBEAT, },
+        { MSG_NEXT_PARAM, },
+    };
+    // returns index of id in deferred_message[] or -1 if not present
+    int8_t deferred_message_offset(const ap_message id) const;
+    // returns index of a message in deferred_message[] which should
+    // be sent (or -1 if none to send at the moment)
+    int8_t deferred_message_to_send_offset();
+    // cache of which deferred message should be sent next:
+    int8_t _next_deferred_message_to_send = -1;
+
+    struct deferred_message_bucket_t {
+        Bitmask ap_message_ids{MSG_LAST};
+        uint16_t interval_ms;
+        uint16_t last_sent_ms; // from AP_HAL::millis16()
+    };
+    deferred_message_bucket_t deferred_message_bucket[5];
+    static const uint8_t no_bucket_to_send = -1;
+    static const ap_message no_message_to_send = (ap_message)-1;
```

Comment:
it's history but we probably should have called the ap_message enum ap_message_id.

Reply:
Some future PR :-)

And (just to get the ball rolling) I'm not sure it shouldn't be `ap_message_t` :-)